File: 2asir/web-deployment/PHP/clase/OBJETOS/Python_POO_Mastery_Kit/04_Logica_Avanzada.py
"""
04_Logica_Avanzada.py

OBJETIVO:
Herencia, Clases Abstractas y Polimorfismo.

DIFERENCIAS:
- Herencia: `class Hija(Padre):` en vez de `extends`.
- Abstractas: Python necesita importar `ABC` y `abstractmethod`.
- Interfaces: En Python no existe `interface`. Se usan clases abstractas puras.
"""

from abc import ABC, abstractmethod

# CLASE ABSTRACTA (Equivalente a abstract class Animal)
class Animal(ABC):
    
    def __init__(self, nombre):
        self.nombre = nombre
        
    def dormir(self):
        return f"{self.nombre} está durmiendo: Zzz..."
    
    # Método abstracto (obliga a definirlo en hijos)
    @abstractmethod
    def moverse(self):
        pass
    
    @abstractmethod
    def hacer_sonido(self):
        pass

# CLASE HIJA 1
class Perro(Animal):
    def hacer_sonido(self):
        return "¡Guau!"
    
    def moverse(self):
        return "corriendo en 4 patas"
    
    def buscar_pelota(self):
        return f"{self.nombre} fue a buscar la pelota."

# CLASE HIJA 2
class Pajaro(Animal):
    def hacer_sonido(self):
        return "¡Pío pío!"
    
    def moverse(self):
        return "volando alto"

## -----------------------------------------
## POLIMORFISMO
## -----------------------------------------

# Python no requiere declarar el tipo de datos (type hinting opcional)
# def presentar_animal(animal: Animal): funcionaría para documentar
def presentar_animal(animal):
    print(f"Aquí tenemos a {animal.dormir()}.")
    print(f"Dice: {animal.hacer_sonido()}")
    print(f"Se mueve: {animal.moverse()}")
    print("---------------------------")

mi_perro = Perro("Firulais")
mi_pajaro = Pajaro("Tweety")

# abstract = Animal("Test") # DARÍA ERROR

presentar_animal(mi_perro)
presentar_animal(mi_pajaro)

print(mi_perro.buscar_pelota())

File: 2asir/web-deployment/PHP/clase/OBJETOS/PHP_POO_Mastery_Kit/04_Logica_Avanzada.php
<?php
/**
 * 04_Logica_Avanzada.php
 * 
 * OBJETIVO:
 * Jerarquías de clases y Polimorfismo. No repetir código usando herencia.
 * 
 * CONCEPTOS:
 * 1. Herencia (extends): Una clase hija hereda todo de la clase padre.
 * 2. Protected: Visible en la clase padre e hijas (pero no fuera).
 * 3. Clases Abstractas: Plantillas que no se pueden instanciar, obligan a las hijas a cumplir reglas.
 * 4. Interfaces: Contratos estrictos que definen QUÉ métodos debe haber.
 */

// INTERFAZ: Define un "contrato". Todo animal DEBE poder hacer sonido.
interface SerVivo
{
    public function hacerSonido();
}

// CLASE ABSTRACTA: Define la base. No existe un "Animal" genérico en la realidad.
abstract class Animal implements SerVivo
{
    protected $nombre; // Las hijas pueden acceder aquí

    public function __construct($nombre)
    {
        $this->nombre = $nombre;
    }

    // Método común para todos
    public function dormir()
    {
        return "{$this->nombre} está durmiendo: Zzz...";
    }

    // Método abstracto: Las hijas ESTÁN OBLIGADAS a definir cómo se mueven
    abstract public function moverse();
}

// CLASE HIJA 1
class Perro extends Animal
{
    public function hacerSonido()
    {
        return "¡Guau!";
    }

    public function moverse()
    {
        return "corriendo en 4 patas";
    }

    public function buscarPelota()
    {
        return "{$this->nombre} fue a buscar la pelota.";
    }
}

// CLASE HIJA 2
class Pajaro extends Animal
{
    public function hacerSonido()
    {
        return "¡Pío pío!";
    }

    public function moverse()
    {
        return "volando alto";
    }
}

## -----------------------------------------
## SECCIÓN DE POLIMORFISMO
## -----------------------------------------

// Función que acepta CUALQUIER Animal (Polimorfismo)
function presentarAnimal(Animal $animal)
{
    echo "Aquí tenemos a {$animal->dormir()}.\n";
    echo "Dice: " . $animal->hacerSonido() . "\n";
    echo "Se mueve: " . $animal->moverse() . "\n";
    echo "---------------------------\n";
}

$miPerro = new Perro("Firulais");
$miPajaro = new Pajaro("Tweety");

// $animalGen = new Animal("Test"); // ERROR: No se puede instanciar abstractas

presentarAnimal($miPerro);
presentarAnimal($miPajaro);

// Método exclusivo del perro
echo $miPerro->buscarPelota() . "\n";

/*
 * NOTA DOCENTE:
 * La herencia permite reutilizar el código de `dormir()`.
 * El polimorfismo permite tratar a Perros y Pájaros como "Animales" genéricos
 * en la función `presentarAnimal`, ahorrando código duplicado.
 */
?>